### PR TITLE
Build the Mode-A knowledge-distillation pipeline

### LINKED ISSUE #414: Build the Mode-A knowledge-distillation pipeline (soft labels + span logits)
## Summary
Section 6.3 Mode A distills the Mode-C teacher into a distilbert/small encoder via KD (soft labels + span logits) -> INT8, and section 6.6 names distillation as the 2026 frontier. OM-038a only commits the tiny_distill.yaml preset (config, no logic) and OM-050/053 consume distillation as undifferentiated XL epic steps. NO task implements the actual KD pipeline: teacher inference to soft labels, span-logit transfer, temperature/alpha mixing, and student-fit loss. This builds that discrete, reusable distillation tooling so the Nano/Tiny student tasks and the SLM epic call one tested code path.

## Scope
- [ ] Add openmed/training/distill.py implementing token-classification KD: load a teacher checkpoint, run inference to per-token soft-label distributions and span logits, and produce a KD loss combining teacher-soft-label KL (temperature-scaled) with hard-label CE and a span-logit transfer term (mixing weight alpha).
- [ ] Support distilling onto a smaller encoder declared by the tiny_distill preset (OM-038a); reuse core/decoding for span repair so the student inherits the teacher's span structure.
- [ ] Emit a distillation report (teacher id, student backbone, temperature, alpha, per-label student-vs-teacher recall delta) feeding the model card evidence block.
- [ ] Honour the recall-first invariant: the report flags any critical-label recall drop vs the teacher so OM-031b's G4-style gate can block it downstream.
- [ ] Tests: KD loss reduces to CE when alpha=0 and temperature=1; span-logit transfer term is computed; distillation report records per-label teacher-vs-student deltas.

## Acceptance criteria
- [ ] openmed/training/distill.py runs teacher->student KD with temperature-scaled soft-label KL, hard-label CE, and span-logit transfer terms.
- [ ] The student backbone is read from the tiny_distill preset; span repair reuses core/decoding.
- [ ] A distillation report records per-label teacher-vs-student recall deltas and flags critical-label drops.
- [ ] KD loss collapses to plain CE at alpha=0, temperature=1 (sanity invariant tested).
- [ ] test suite green: .venv/bin/python -m pytest tests/ -q

## Out of scope
- Producing the actual Nano/Tiny distilled artifacts (separate Nano task / training epics).
- The recipe package and tiny_distill preset (OM-038a).
- Quantization to INT8 (covered by export/quant work).

## Files
- openmed/training/distill.py
- openmed/training/configs/tiny_distill.yaml
- tests/unit/training/test_distill.py

---
Task: OM-157  ·  Milestone: v1.7  ·  Priority: P1  ·  Size: L
Depends on: OM-038a  ·  Blocks: —
Roadmap: OPENMED_V2_UNIFIED_ROADMAP.md section 6.3 (Mode A), 6.6
Spec: PLANS/V2/EXECUTION/tasks/OM-157.md

### PR DESCRIPTION
**Pull request by maziyarpanahi**
# Pull Request

## Description
Closes #414.

Adds Mode-A token-classification distillation support with teacher soft-label targets, temperature-scaled KL, hard-label cross-entropy, span-logit transfer, shared BIOES span repair, tiny preset student-backbone lookup, and recall-delta reporting for model-card evidence.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [x] Test addition/improvement

## Changes Made
- Added `openmed.training.distill` with KD loss components, teacher target helpers, repaired-span decoding, and report dataclasses.
- Exported the distillation API through the existing lazy `openmed.training` surface.
- Added offline unit coverage for CE collapse, span-logit transfer, soft labels, preset lookup, span repair, and critical-label recall drops.

## Testing
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have tested this change with different models/inputs

Commands run:
- `.venv/bin/python -m pytest tests/ -q`
- `make format`
- `make lint`
- `make format-check`

## Documentation
- [ ] I have updated the documentation accordingly
- [x] I have added docstrings to new functions/classes
- [ ] I have updated the CHANGELOG.md

## Code Quality
- [x] I ran `make format`, `make lint`, and `make format-check`
- [ ] For Swift/OpenMedKit changes, I ran `make format-swift` and `make lint-swift`
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

## Dependencies
- [x] I have not added any new dependencies
- [ ] OR I have added new dependencies and they are justified because: ____

## Checklist
- [x] I have read the contributing guidelines
- [x] My commits have clear, descriptive messages
- [x] I have squashed/organized my commits appropriately

## Related Issues
Closes #414

## Screenshots/Examples
Not applicable.

### CODE REVIEW

**Comment by maziyarpanahi** [comment]
Thanks for the PR. I reviewed the issue, PR, and comments, then updated the branch against current `master` and added a maintainer follow-up commit.

What I changed:

- made the `alpha=0` KD sanity path actually runnable without teacher tensors, while still requiring teacher outputs when `alpha > 0`
- filtered zero-offset special tokens before BIOES/Viterbi repair so specials cannot influence repaired teacher spans
- added deterministic JSON writing for `DistillationReport`
- wired distillation evidence into generated model cards when a manifest row includes the report block
- added coverage for CE-only loss, teacher-output validation, special-token filtering, report JSON, model-card evidence, and the optional `ModeADistillationPipeline.teacher_targets()` / `student_loss()` path with offline fake models

Validation run:

- `/Users/maziyar/Developer/openmed/.venv/bin/python -m pytest tests/unit/training/test_distill.py tests/unit/core/test_model_card.py -q`
- `/Users/maziyar/Developer/openmed/.venv/bin/pre-commit run --files openmed/training/distill.py openmed/core/model_card.py tests/unit/training/test_distill.py tests/unit/core/test_model_card.py`
- GitHub CI is green across repo-policy, lint, security, secret-scan, test matrix, and build.

The issue labels were already present on the PR. The branch is current with `master`, conflict-free, and mergeable from my side.
